autofaiss/datasets/readers/local_iterators.py

```diff
@@ -52,58 +61,48 @@ def read_embeddings_local(local_embeddings_path: str, stack_input: int = 1, verb
-    """
-    Iterate over numpy array files that match the reg ex pattern and yield their content.
-    It is possible to iterate over the stacked content of several arrays.
-
-    Parameters
-    ----------
-    local_embeddings_path : str
-        Path on local disk of arrays in numpy format.
-    stack_input : int (default 1)
-        Number of arrays that should be stacked at each iterations.
-        This parameter is useful when working with many small files.
-    verbose : bool
-        Print detailed informations if set to True
-
-    Returns
-    -------
-    arrays_iterator : Iterator[np.ndarray]
-        An iterator over batchs of stacked arrays.
-    """
-
-    assert stack_input > 0
+    if batch_size is None:
+        batch_size = first_vector_count
 
     reg_exp = re.compile(reg_exp_pattern)
 
-    filenames = os.walk(local_path).__next__()[2]
+    filenames = os.walk(local_embeddings_path).__next__()[2]
     filenames = [filename for filename in filenames if reg_exp.match(filename)]
     filenames.sort()
-    embeddings_stack: List[np.ndarray] = []
+    embeddings_batch = None
+    nb_emb_in_batch = 0
 
-    iterator = enumerate(filenames)
+    iterator = filenames
     if verbose:
         iterator = tq(list(iterator))
 
-    for file_number, file_name in iterator:
-
-        if embeddings_stack and (file_number % stack_input == 0):
-            yield np.concatenate(embeddings_stack)
-            embeddings_stack = []
-
-        try:
-            emb = np.load(f"{local_path}/{file_name}")
-            if emb.dtype != "float32":
-                emb = emb.astype("float32")
-            embeddings_stack.append(emb)
-        except Exception as e:  # pylint: disable=broad-except
-            print(e)
-
-    if embeddings_stack:
-        yield np.concatenate(embeddings_stack).astype(np.float32)
+    for file_name in iterator:
+        emb = np.load(f"{local_embeddings_path}/{file_name}", mmap_mode="r")
+        vec_size = emb.shape[0]
+        current_emb_index = 0
+        while True:
+            left_in_emb = vec_size - current_emb_index
+            remaining_to_add = max(batch_size - nb_emb_in_batch, 0)
+            adding = min(remaining_to_add, left_in_emb)
+            additional = max(left_in_emb - adding, 0)
+            if embeddings_batch is None:
+                embeddings_batch = np.empty((batch_size, dim), "float32")
+            embeddings_batch[nb_emb_in_batch : (nb_emb_in_batch + adding), :] = emb[
+                current_emb_index : (current_emb_index + adding), :
+            ]
+            nb_emb_in_batch += adding
+            current_emb_index += adding
+            if nb_emb_in_batch == batch_size:
+                yield embeddings_batch
+                nb_emb_in_batch = 0
+                embeddings_batch = None
+            if additional == 0:
+                break
+
+    if nb_emb_in_batch > 0 and embeddings_batch is not None:
+        yield embeddings_batch[:nb_emb_in_batch]
```

Comment:
This is not necessary because the size is always < nb_emb_in_batch
Keeping the code like this might mislead a reader who would think we discard some vectors?

Reply:
that's the last batch which may be partial yes, I agree adding that in comment can help